### PR TITLE
Move to JUnit 5

### PR DESCRIPTION
**Pull request by avandeursen**
Experimental branch with JUnit 5 version of JPacman.

This move also implies a systematic adoption of AssertJ (and dropping of hamcrest).

All seems to working fine -- still exploring the impact on some of the exercises.

Feedback welcome.

(Do not merge yet).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -45,16 +45,21 @@
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
-		<cucumber.version>1.2.4</cucumber.version>
-		<assertj.version>3.4.1</assertj.version>
-		<junit.version>4.12</junit.version>
-		<mockito.version>2.0.50-beta</mockito.version>
+        <junit.jupiter.version>5.0.0-M3</junit.jupiter.version>
```
Comment (TimvdLippe) [inline pom.xml]: Nit: indentation seems off here

File: src/test/java/nl/tudelft/jpacman/LauncherSmokeTest.java [hunk 2]
```diff
@@ -55,37 +52,37 @@ public void tearDown() {
...
         game.start();
-        assertTrue(game.isInProgress());
-        assertEquals(0, player.getScore());
+        assertThat(game.isInProgress()).isTrue();
+        assertThat(player.getScore()).isZero();
 
         // get points
         game.move(player, Direction.EAST);
-        assertEquals(10, player.getScore());
+        assertThat(player.getScore()).isEqualTo(10);
```
Comment (TimvdLippe) [inline src/test/java/nl/tudelft/jpacman/LauncherSmokeTest.java]: I really love that tests now read like stories. Big :+1: for AssertJ

File: src/test/java/nl/tudelft/jpacman/board/BoardFactoryTest.java [hunk 3]
```diff
@@ -18,79 +20,75 @@
...
-		factory.createBoard(grid);
-		assertEquals(s, s.getSquareAt(Direction.NORTH));
-		assertEquals(s, s.getSquareAt(Direction.SOUTH));
-		assertEquals(s, s.getSquareAt(Direction.WEST));
-		assertEquals(s, s.getSquareAt(Direction.EAST));
+		// Create a round world with just one cell.
+		factory.createBoard(new Square[][]{{s1}});
+
+		// All directions from that single cell point back to that cell.
+		Arrays.stream(Direction.values()).forEach((dir) ->
```
Comment (TimvdLippe) [inline src/test/java/nl/tudelft/jpacman/board/BoardFactoryTest.java]: I am not sure how well the error is reported if 1 of them is not true. Maybe we can use a `groupedAssertion` here. See  http://junit.org/junit5/docs/current/user-guide/#writing-tests-assertions for an example with `assertAll`

Comment (LiamClark) [inline src/test/java/nl/tudelft/jpacman/board/BoardFactoryTest.java]: Could also be:

assertThat(Arrays.stream(Direction.values()).map(s1::getSquareAt)).containsOnly(s1);

File: src/test/java/nl/tudelft/jpacman/board/BoardTest.java [hunk 4]
```diff
@@ -1,82 +1,70 @@
...
+    public void verifyHeight() {
+        assertThat(board.getHeight()).isEqualTo(MAX_HEIGHT);
+    }
+
+    /**
+     * Test at various places that the board returns the correct square.
+     * @return Tests for the various places.
+     */
+    @TestFactory
+    public Iterable<DynamicTest> verifySquares() {
```
Comment (LiamClark) [inline src/test/java/nl/tudelft/jpacman/board/BoardTest.java]: This could use the new parameterized support in m4

Comment (avandeursen) [inline src/test/java/nl/tudelft/jpacman/board/BoardTest.java]: Yes. Will look at that.

File: pom.xml [hunk 5]
```diff
@@ -45,11 +45,11 @@
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
-        <junit.jupiter.version>5.0.0-M3</junit.jupiter.version>
-        <junit.platform.version>1.0.0-M3</junit.platform.version>
+        <junit.jupiter.version>5.0.0-M4</junit.jupiter.version>
+        <junit.platform.version>1.0.0-M4</junit.platform.version>
```
Comment (avandeursen) [inline pom.xml]: This change might have been a bit too fast, as it breaks testing in IntelliJ at the moment. But that will be fixed soon I assume.

Comment (LiamClark) [inline pom.xml]: Indeed, Some issue on junit5 repo also mentions this, since the current intelij support is still based on M3

Comment (avandeursen) [inline pom.xml]: https://youtrack.jetbrains.com/issue/IDEA-170817

**Comment by TimvdLippe** [comment]
So we are going to disable this bot? :joy:

**Comment by avandeursen** [comment]
Did anyone enable the Codacy bot?

The rules of the bot should be checkable local as well, through maven. 

Are unsolicited bot remarks in agreement with the GitHub terms of service?

**Comment by TimvdLippe** [comment]
Maybe @JWGmeligMeyling or @michaeldejong enabled it?

**Comment by JWGmeligMeyling** [comment]
No I didn't, pretty sure Michael didn't too.


Jan-Willem

**Comment by avandeursen** [comment]
I propose to merge this in, and commit to using JUnit 5.

**Comment by avandeursen** [comment]
OK, I will block this tool as per https://github.com/codacy-bot/how-to-unsubscribe

**Pull request merged**